openmm: init at 7.7.0

sheepforce opened the pull request:
###### Description of changes

Part of the ongoing effort to upstream packages from the [NixOS-Qchem overlay](https://github.com/markuskowa/NixOS-QChem) (https://github.com/markuskowa/NixOS-QChem/issues/146). This adds the [OpenMM](https://openmm.org/) package for classical molecular dynamics.

The package works when explicitly setting `PYTHONPATH` but cannot be imported otherwise. This is also the reason why the `pythonImportsCheck` is currently commented out. I couldn't figure out what the problem is and appreciate any idea :)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

azahi commented on pkgs/development/python-modules/openmm/default.nix:
```diff
@@ -0,0 +1,87 @@
+{ buildPythonPackage, lib, fetchFromGitHub, cmake, clang, gfortran, fftwSinglePrec
+, doxygen, ocl-icd, opencl-headers, swig, python, cython, numpy
+, cudatoolkit, enableCuda ? false
+} :
```
```suggestion
{ buildPythonPackage
, lib
, fetchFromGitHub
, cmake
, clang
, gfortran
, fftwSinglePrec
, doxygen
, ocl-icd
, opencl-headers
, swig
, python
, cython
, numpy
, enableCuda ? false, cudatoolkit
}:
```

sheepforce replied:
Hm why? :slightly_smiling_face:  nixfmt agrees with this format and many other packages use it.

azahi replied:
Well, other packages also use this format, plus it's easier to manage diffs with something like this. Speaking of formatting, AFIAK there are no standard formatters for nixpkgs. There's a `nixpkgs-fmt` tho, which I guess is more geared towards nixpkgs.

sheepforce replied:
You are right about the diffs. The nix tooling gets me every time ... However, has the suggested format now as by nixpkgs-fmt

azahi commented on pkgs/development/python-modules/openmm/default.nix:
```diff
@@ -0,0 +1,87 @@
+    fftwSinglePrec
+    ocl-icd
+    opencl-headers
+  ] ++ lib.optional enableCuda cudatoolkit
+  ;
```
```suggestion
  ] ++ lib.optional enableCuda cudatoolkit;
```

sheepforce replied:
Run nixfmt on it again, changed.

azahi commented on pkgs/development/python-modules/openmm/default.nix:
```diff
@@ -0,0 +1,87 @@
+    python setup.py build && python setup.py install --prefix=$out
+  '';
+
+  # Needs to explicitly set PYTHONPATH when using.
+  #pythonImportsCheck = [ "openmm.app" ];
```
What's this for?

sheepforce replied:
See my initial message in the PR.

>The package works when explicitly setting PYTHONPATH but cannot be imported otherwise. This is also the reason why the pythonImportsCheck is currently commented out. I couldn't figure out what the problem is and appreciate any idea :)

azahi replied:
Hmm, so basically we break either binary usage or importing?

sheepforce replied:
Could you elaborate how these two relate? Why does binary usage break importing or vice versa? I feel I am missing some knowledge here.

azahi replied:
I'm as confused here as you are. I guess we need some more eyes to look at this.

RaitoBezarius replied:
Is there any confusion on the module name due to the presence of a folder in the working directory of something?

sheepforce replied:
> Is there any confusion on the module name due to the presence of a folder in the working directory of something?

No I don't think so. Happens even from an empty directory and the directory structure of the result looks perfectly normal:

```
result/lib/
├── libOpenMMAmoeba.so
├── libOpenMMDrude.so
├── libOpenMMRPMD.so
├── libOpenMM.so -> libOpenMM.so.7.7
├── libOpenMM.so.7.7
├── plugins
│   ├── libOpenMMAmoebaOpenCL.so
│   ├── libOpenMMAmoebaReference.so
│   ├── libOpenMMCPU.so
│   ├── libOpenMMDrudeOpenCL.so
│   ├── libOpenMMDrudeReference.so
│   ├── libOpenMMOpenCL.so
│   ├── libOpenMMPME.so
│   ├── libOpenMMRPMDOpenCL.so
│   └── libOpenMMRPMDReference.so
└── python3.10
    └── site-packages
        ├── openmm
        ├── OpenMM-7.7.0-py3.10.egg-info
        └── simtk
``` 



> If the python package can't be imported it is likely broken and does not work correct.

Works perfectly as expected when one sets the PYTHONPATH. That is not the problem.

RaitoBezarius replied:
Alright, this requires investigation then.

sheepforce replied:
OK, I am building now with stdenv.mkDerivation and convert with `toPythonModule` afterwards. Looks more clean. I can also import openmm from a nix-shell. A test from within the derivation somehow seems to suffer from an import of a swig-generated interface. 
`result/lib/python3.10/site-packages/openmm/_openmm*.so` is not found before the derivation is finalised. However, the finished derivation can be imported from Python without problems.

azahi commented on pkgs/development/python-modules/openmm/default.nix:
```diff
@@ -0,0 +1,87 @@
+  ;
+
+  propagatedBuildInputs = [
+    python
```
Maybe it would be better to explicitly specify that this is `python2`?

sheepforce replied:
is gone

azahi commented on pkgs/development/python-modules/openmm/default.nix:
```diff
@@ -0,0 +1,87 @@
+  dontUseSetuptoolsCheck = true;
+
+  cmakeFlags = [
+    "-DCMAKE_C_COMPILER=${clang}/bin/clang"
+    "-DCMAKE_CXX_COMPILER=${clang}/bin/clang++"
```
Is there are a reason to use clang?

sheepforce replied:
There is: http://docs.openmm.org/7.1.0/userguide/library.html#installing-on-linux-and-mac

Basically performance on CPU is much better with Clang and I would argue that for a package like OpenMM this is of importance. I will make Clang optional but enable by default. :slightly_smiling_face:

azahi commented on pkgs/development/python-modules/openmm/default.nix:
```diff
@@ -0,0 +1,87 @@
+    "-DOPENMM_BUILD_C_AND_FORTRAN_WRAPPERS=ON"
+    "-DOPENMM_BUILD_DRUDE_PLUGIN=ON"
+    "-DOPENMM_BUILD_PME_PLUGIN=ON"
+    "-DOPENMM_BUILD_PYTHON_WRAPPERS=ON"
```
I'm assuming this provides a Python 2 API. Since Python 2 is deprecated, can we hide that behind an option?

sheepforce replied:
Is purely python3 now.

azahi commented on pkgs/development/python-modules/openmm/default.nix:
```diff
@@ -0,0 +1,87 @@
+    "-DCMAKE_LIBRARY_PATH=${cudatoolkit}/lib64/stubs"
+  ];
+
+  enableParallelBuilding = true;
+
```
This is the default.
```suggestion
```

azahi commented on pkgs/development/python-modules/openmm/default.nix:
```diff
@@ -0,0 +1,73 @@
+  #pythonImportsCheck = [ "openmm.app" ];
+
+  meta = with lib; {
+    description =
+      "Toolkit for molecular simulation using high performance GPU code";
```
This is why I don't like `nixfmt`. :^)
```suggestion
    description = "Toolkit for molecular simulation using high performance GPU code";
```

sheepforce replied:
Granted, this is weird. Changed it back :)

SuperSandro2000 commented on pkgs/development/python-modules/openmm/default.nix:
```diff
@@ -0,0 +1,86 @@
+    cd python
+    export OPENMM_INCLUDE_PATH=$out/include
+    export OPENMM_LIB_PATH=$out/lib
+    python setup.py build && python setup.py install --prefix=$out
```
```suggestion
    ${python.interpreter} setup.py build
    ${python.interpreter} setup.py install --prefix=$out
```
Can't we use the standard buildPythonPackage phases?

sheepforce replied:
No, the C++ stuff is always built with CMake, see the OpenMM manual: http://docs.openmm.org/latest/userguide/library/02_compiling.html

sheepforce replied:
Formatting has been changed but `${python.interpreter}` does not exist, `python` itself is not passed as we are already in `buildPythonPackage`.

sheepforce replied:
Has been changed. I am building now as a C++ package with python bindings and call `toPythonModule` afterwards.

SuperSandro2000 commented on pkgs/development/python-modules/openmm/default.nix:
```diff
@@ -0,0 +1,86 @@
+, numpy
+, cudatoolkit
+, clang
+, enableClang ? true
+, enableCuda ? false
```
We shouldn't give pythonPackages packages such options that the user needs to change.

sheepforce replied:
There is no **need** to change. I think the defaults are reasonable. And such config options are used for the tensorflow and pytorch packages as well?

sheepforce replied:
OpenMM and its C++, C and Fortran bindings now live in all-packages. Flags to control whether to enable CUDA (false by default), OpenCL (false by default), CLang (true by default) and Python bindings are set. The defaults are reasonable IMHO and pass all tests.

SuperSandro2000 commented on pkgs/development/libraries/science/chemistry/openmm/default.nix:
```diff
@@ -0,0 +1,90 @@
+    hash = "sha256-2PYUGTMVQ5qVDeeABrwR45U3JIgo2xMXKlD6da7y3Dw=";
+  };
+
+  # "This test is stochastic and may occassionally fail". It does.
+  postPatch = ''
+    rm \
+      platforms/*/tests/Test*BrownianIntegrator.* \
+      serialization/tests/TestSerializeIntegrator.cpp
```
and another one...
```
exception: Assertion failure at TestLangevinIntegrator.h:123.  Expected 9.97736, found 10.6743 (This test is stochastic and may occasionally fail)
```

sheepforce replied:
Meh, that is uncool. Probably half of the test suite could have stochastic failures. This is somehow how molecular dynamics works. I have a hard time figuring out which of them are expected to stochastically fail, though. Let's hope this was the last one.

sheepforce committed:
openmm: init at 7.7.0


openmm formatting and inputs


openmm: clang optional


openmm: formatting


openmm: formatting


openmm: formatting


openmm: formatting


openmm: fix


openmm: disable more tests


openmm: disable more tests


openmm: disable more stochastic tests

markuskowa reviewed (approved):
Looks good to me. Let's go with it for now and make adjustments later if the tests cause more trouble.
